crossbeam/algorithm/beam_search.py

```diff
@@ -43,36 +72,57 @@ def beam_search(beam_steps, k, choice_embed, init_embed, score_model, device, ch
     prefix_scores: scores for each arg-list
   """
   cur_state = score_model.get_init_state(init_embed, batch_size=1)
-  num_choices = choice_embed.shape[0]
   prefix_scores = torch.zeros(1).to(device)
   arg_choices = torch.LongTensor([[]]).to(device)
-  for step in range(beam_steps):
-    scores = score_model.step_score(cur_state, choice_embed)  # result in a score matrix of size (N-state, N-choice)
-    joint_scores = prefix_scores.unsqueeze(1) + scores # broadcast over columns    
+
+  # select args
+  for step in range(arity):
     if choice_masks is not None and len(choice_masks):
       choice_mask = choice_masks[step][1].view(1, -1).float()
-      joint_scores = joint_scores * choice_mask + (1 - choice_mask) * N_INF
-    joint_scores = joint_scores.view(-1)
-    cur_k = joint_scores.shape[0] if k > joint_scores.shape[0] else k
-    if is_stochastic:
-      prob = torch.softmax(joint_scores, dim=0)
-      arg_selected = torch.multinomial(prob, cur_k)
-      prefix_scores = joint_scores[arg_selected]
-      prefix_scores, idx_sorted = torch.sort(prefix_scores)
-      arg_selected = arg_selected[idx_sorted]
     else:
-      prefix_scores, arg_selected = torch.topk(joint_scores, cur_k)
-    prev_index = torch.div(arg_selected, num_choices, rounding_mode='floor')
-    op_choice = arg_selected % num_choices
-
-    prev_state = score_model.state_select(cur_state, prev_index)
-    cur_op_embed = choice_embed[op_choice]
-    cur_state = score_model.step_state(prev_state, cur_op_embed)
-
+      choice_mask = None
+    cur_state, prev_index, op_choice, prefix_scores = _beam_step(score_model, k, cur_state,
+                                                                 value_embed, prefix_scores,
+                                                                 choice_mask=choice_mask,
+                                                                 is_stochastic=is_stochastic)
     new_arg_choices = arg_choices[prev_index]
     new_arg_choices = torch.cat((new_arg_choices, op_choice.unsqueeze(1)), axis=1)
     arg_choices = new_arg_choices
-  return arg_choices, prefix_scores
+
+  # bind free_var/bond_var for each arg
+  for step in range(arity):
+    for inner_step in range(MAX_NUM_BOUND_VARS + MAX_NUM_FREE_VARS):
+      cur_args = arg_choices[:, step].detach().cpu().numpy()
+      num_arg_vars = []
+      for i in cur_args:
+        arg = values[i]
+        num_var = 0 if isinstance(arg, value_module.FreeVariable) else arg.num_free_variables
+        num_arg_vars.append(num_var)
+      stop_indices = torch.LongTensor([i for i, x in enumerate(num_arg_vars) if x <= inner_step]).to(device)
+      cur_beam_size = k - stop_indices.shape[0]
+
+      if cur_beam_size == 0:
+        invalid_args = torch.zeros(arg_choices.shape[0], special_var_embed.shape[0] - inner_step).to(arg_choices) - 1
```

Comment:
`special_var_embed.shape[0]` is the number of distinct free and bound variables. what we want here is the max number of variables predicted for one argument. these two are currently the same by chance but aren't conceptually the same

Reply:
you are right, with your above suggested fixes these two are no longer equal.